rfcs/20200624-pluggable-device-for-tensorflow.md

```diff
@@ -0,0 +1,251 @@
+<div align="center">
+<img src=20200624-pluggable-device-for-tensorflow/gpu_example.png>
+</div>
+
+### Device Discovery
+
+Upon initialization of TensorFlow, it uses platform independent `LoadLibrary()` to load the dynamic library. The plugin library should be installed to default plugin directory "…python_dir.../site-packages/tensorflow-plugins". The modular tensorflow [RFC](https://github.com/tensorflow/community/pull/77) describes the process of loading plugins. 
+
+During the plugin library initialization, it calls the `SE_ReigsterPlatform()` API to register the stream executor platform (`SE_Platform` struct) to TensorFlow proper. The `SE_ReigsterPlatform()` API is a callback API, part of StreamExecutor C API, which passes necessary information to TensorFlow proper to instantiate a stream executor platform ([se::platform](https://github.com/tensorflow/tensorflow/blob/cb32cf0f0160d1f582787119d0480de3ba8b9b53/tensorflow/stream_executor/platform.h#L93) class) and register to a global object [se::MultiPlatformManager](https://github.com/tensorflow/tensorflow/blob/cb32cf0f0160d1f582787119d0480de3ba8b9b53/tensorflow/stream_executor/multi_platform_manager.h#L82). 
+The stream executor platform must be registered with the name "PluggableDevice".  
+See below code which is an example of registering a PluggableDevice platform with StreamExecutor C API:
+```cpp
+void RegisterPluggableDevicePlatform() {
+  static plugin_id_value = 123;
+  SE_PlatformId id;
+  id.id = &plugin_id_value;
+  int visible_device_count = get_plugin_device_count;
+  SE_Platform* custom_platform = SE_NewPlatform(
+     id, visible_device_count,
+     create_device, create_stream_executor,
+     delete_device, delete_stream_executor);
+  TF_Status* status = TF_NewStatus();
+  std::string name = "PluggableDevice";
+  SE_RegisterPlatform(
+     name.c_str(), name.size(),
+     custom_platform,
+     status);
+}
+
+```
+Use static initialization to register the new platform:
+```cpp
+static bool IsPluggableDevicePlatformRegistered = []() {
+ RegisterPluggablePlatform();
+ return true;
+}();
+
+```
+
+### Device Creation
+
+`PluggableDeviceFactory` is introduced to create the `PluggableDevice`, following the [LocalDevice](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/local_device.h) design pattern. To support existing GPU programs running on a new device without user changing the code, PluggableDeviceFactory is registered as "GPU" name and given higher priority than the default GPU device.  
+&emsp;&emsp;`REGISTER_LOCAL_DEVICE_FACTORY("GPU",PluggableDeviceFactory, 220); // plugged GPU`  
+&emsp;&emsp;`REGISTER_LOCAL_DEVICE_FACTORY("GPU", GPUDeviceFactory, 210);//default GPU`  
+For those vendor who don't want to use "GPU" name, it's optional to register a new device name. For example:   
+&emsp;&emsp;`REGISTER_LOCAL_DEVICE_FACTORY("Third-party device",PluggableDeviceFactory, 230); // plugged third party device`
```

Comment:
I think it is a good idea. it makes sense that the device name(front end) is registered from plugin. Will you change the C API? If if is defined, I will change it.Thanks.
And let me confirm more about the names in the framework, now we have 3 names need to be dealt with:
    device name(front-end): use to register Device Factory which maps to the python front end name, can be dynamic registered
    platform type(back-end): a name used to connect with Device and StreamExecutor Platform, can be dynamic registered(optional)
    device type(back-end): DEVICE_PLUGGABLE(defined in framework/types.cc) , a name used to register kernel and deal with backend divergent, should be a static name,  for example : https://github.com/tensorflow/tensorflow/blob/cbb94efa589fc1d1b34948b9ca03829fe8613d12/tensorflow/core/framework/op_kernel.cc#L145

Reply:
I updated StreamExecutor RFC to include that snippet now: https://github.com/tensorflow/community/pull/257
Hm I think device and platform name can be the same (for e.g. CUDA). Device type needs to be set separately, for e.g. GPU.

Reply:
@annarev Thanks for the update!  
   do you prefer the device_type(counter-part of `DEVICE_GPU`) to be dynamically registered from plugin through `SE_RegisterPlatform` or statically defined in Tensorflow Proper? Since it is used frequently in current tensorflow stack, I think this may need more discussions:
**static definition**:
  Tensorflow proper defines DEVICE_PLUGGABLE to hard-code to branch in some common code(such as [ProcessMemoryTypes](https://github.com/tensorflow/tensorflow/blob/d93971b09e699f646f773552d53cf3caad28bdea/tensorflow/core/common_runtime/memory_types.cc#L51)), and plugin  use this string to register the kernels.
  **advantage**: the branch code may be simple:
   ```
    if (device_type != DEVICE_GPU && device_type != DEVICE_SYCL && **device_type != DEVICE_PLUGGABLE**) {
      // On non-GPU and non-SYCL devices, HOST_MEMORY and DEVICE_MEMORY are always
      // compatible.
      return Status::OK();
    }
   ```
  **disadvantage**: only one plugin can be loaded at the same time, multiple plugin will have kernel registration conflict issue.
**dynamic registration""
    the device type may also be registered through `SE_RegisterPlatform`
    **advantage**:
     different plugin can have different device type definition so kernel registration conflict can be resolved.
   **disadvantage**
     branch code may be more complicated. we may need a global pool to store all the device type registered from plugin, and a util function to check the device type:
     ```
      if (device_type != DEVICE_GPU && device_type != DEVICE_SYCL && !**IsPluggedType(device_type)**) {
       // On non-GPU and non-SYCL devices, HOST_MEMORY and DEVICE_MEMORY are always
       // compatible.
       return Status::OK();
     }
     ```

Reply:
Thanks, do you expect the device_type ( attribute of class Device: const DeviceAttributes device_attributes_) such as (DEVICE_GPU, DEVICE_CPU) also be registered from plugin through this API? if it is defined in Tensorflow Proper, then all the plugin use the same name to register its kernel and  kernel registration will get conflict when multiple plugins are loaded.

> I updated StreamExecutor RFC to include that snippet now: #257
> Hm I think device and platform name can be the same (for e.g. CUDA). Device type needs to be set separately, for e.g. GPU.